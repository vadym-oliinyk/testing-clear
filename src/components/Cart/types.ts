export interface CartProps {
  products: FormattedProduct[];
  onAmountChange: (id: number, amount: number) => void;
}
