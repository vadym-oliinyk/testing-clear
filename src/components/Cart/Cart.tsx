import React from 'react';
import Grid from '@material-ui/core/Grid';

import getTotalPrice from '../../utils/products/getTotalPrice';

import CartCard from './components/CartCard';
import EmptyCart from './components/EmptyCart';
import CartFooter from './components/CartFooter';
import { CartProps } from './types';

export default function Cart({
  products,
  onAmountChange,
}: CartProps): JSX.Element {
  const totalPrice = getTotalPrice(products);

  if (!products.length) {
    return <EmptyCart />;
  }

  return (
    <>
      <Grid container spacing={2} justifyContent="center">
        {products.map(({ id, image, title, price, inCartAmount }) => (
          <Grid item xs={10} key={id}>
            <CartCard
              image={image}
              title={title}
              price={price}
              inCartAmount={inCartAmount}
              onLess={() => onAmountChange(id, inCartAmount - 1)}
              onMore={() => onAmountChange(id, inCartAmount + 1)}
            />
          </Grid>
        ))}

        <CartFooter totalPrice={totalPrice} />
      </Grid>
    </>
  );
}
