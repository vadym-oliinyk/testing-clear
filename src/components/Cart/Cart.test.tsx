import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import Cart from './Cart';

const MOCKED_PRODUCTS: FormattedProduct[] = [
  {
    id: 1,
    image: 'https://test.jpg',
    price: 300,
    title: 'Some test product',
    inCartAmount: 1,
  },
  {
    id: 2,
    image: 'https://test.jpg',
    price: 200,
    title: 'Some test product',
    inCartAmount: 2,
  },
  {
    id: 3,
    image: 'https://test.jpg',
    price: 300,
    title: 'Some test product',
    inCartAmount: 1,
  },
];

it('should match snapshot', () => {
  const { container } = render(
    <Cart products={MOCKED_PRODUCTS} onAmountChange={() => {}} />,
  );

  expect(container).toMatchSnapshot();
});

it('should call onAmountChange with right id and new amount', () => {
  const onAmountChangeCallback = jest.fn();
  render(
    <Cart products={MOCKED_PRODUCTS} onAmountChange={onAmountChangeCallback} />,
  );
  const firstProductMoreBtn = screen.getAllByTestId('cart-card-more')[0];
  const secondProductLessBtn = screen.getAllByTestId('cart-card-less')[1];

  fireEvent.click(firstProductMoreBtn);
  fireEvent.click(secondProductLessBtn);

  expect(onAmountChangeCallback.mock.calls).toEqual([
    [1, 2],
    [2, 1],
  ]);
});
