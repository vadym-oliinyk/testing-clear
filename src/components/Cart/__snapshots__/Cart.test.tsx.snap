// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-12n0uxm-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10 css-1uzrzkv-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-imy9bx-MuiPaper-root-MuiCard-root"
      >
        <img
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-m1dg97-MuiCardMedia-root"
          data-testid="cart-card-image"
          src="https://test.jpg"
        />
        <div
          class="MuiBox-root css-j7qwjs"
        >
          <div
            class="MuiCardContent-root css-hrc8z3-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
              data-testid="cart-card-title"
            >
              Some test product
            </div>
            <div
              class="MuiTypography-root MuiTypography-subtitle1 css-1786zc4-MuiTypography-root"
              data-testid="cart-card-price"
            >
              300
               $
            </div>
          </div>
          <div
            class="MuiBox-root css-1ndgitk"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-testid="cart-card-less"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="RemoveIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13H5v-2h14v2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <div
              class="MuiTypography-root MuiTypography-h5 css-41an0y-MuiTypography-root"
              data-testid="cart-card-amount"
            >
              1
            </div>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-testid="cart-card-more"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="AddIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10 css-1uzrzkv-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-imy9bx-MuiPaper-root-MuiCard-root"
      >
        <img
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-m1dg97-MuiCardMedia-root"
          data-testid="cart-card-image"
          src="https://test.jpg"
        />
        <div
          class="MuiBox-root css-j7qwjs"
        >
          <div
            class="MuiCardContent-root css-hrc8z3-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
              data-testid="cart-card-title"
            >
              Some test product
            </div>
            <div
              class="MuiTypography-root MuiTypography-subtitle1 css-1786zc4-MuiTypography-root"
              data-testid="cart-card-price"
            >
              200
               $
            </div>
          </div>
          <div
            class="MuiBox-root css-1ndgitk"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-testid="cart-card-less"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="RemoveIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13H5v-2h14v2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <div
              class="MuiTypography-root MuiTypography-h5 css-41an0y-MuiTypography-root"
              data-testid="cart-card-amount"
            >
              2
            </div>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-testid="cart-card-more"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="AddIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10 css-1uzrzkv-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-imy9bx-MuiPaper-root-MuiCard-root"
      >
        <img
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-m1dg97-MuiCardMedia-root"
          data-testid="cart-card-image"
          src="https://test.jpg"
        />
        <div
          class="MuiBox-root css-j7qwjs"
        >
          <div
            class="MuiCardContent-root css-hrc8z3-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
              data-testid="cart-card-title"
            >
              Some test product
            </div>
            <div
              class="MuiTypography-root MuiTypography-subtitle1 css-1786zc4-MuiTypography-root"
              data-testid="cart-card-price"
            >
              300
               $
            </div>
          </div>
          <div
            class="MuiBox-root css-1ndgitk"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-testid="cart-card-less"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="RemoveIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13H5v-2h14v2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <div
              class="MuiTypography-root MuiTypography-h5 css-41an0y-MuiTypography-root"
              data-testid="cart-card-amount"
            >
              1
            </div>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-testid="cart-card-more"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="AddIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10 css-1nfw97f-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-e631gt-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <span
            class="MuiTypography-root MuiTypography-h6 css-gi7hmv-MuiTypography-root"
          >
            Total:
          </span>
          <span
            class="MuiTypography-root MuiTypography-h6 css-oblp1g-MuiTypography-root"
            data-testid="total-price"
          >
            1000
             $
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should match snapshot for empty page 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-e631gt-MuiPaper-root-MuiCard-root"
  >
    <span
      class="MuiTypography-root MuiTypography-h6 css-gi7hmv-MuiTypography-root"
    >
      No products in cart
    </span>
  </div>
</div>
`;
